src/crimson/osd/osd_operation.h

```diff
@@ -0,0 +1,218 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include <seastar/core/shared_mutex.hh>
+#include <seastar/core/future.hh>
+
+#include <vector>
+#include <set>
+#include <boost/intrusive/list.hpp>
+#include <boost/intrusive_ptr.hpp>
+#include <boost/smart_ptr/intrusive_ref_counter.hpp>
+
+#include "common/Formatter.h"
+
+namespace ceph::osd {
+
+enum class OperationTypeCode {
+  client_request = 0,
+  peering_event,
+  compound_peering_request,
+  last_op
+};
+
+static constexpr const char* const OP_NAMES[] = {
+  "client_write",
+  "peering_event",
+  "compound_peering_request"
+};
+
+class OperationRegistry;
+
+using registry_hook_t = boost::intrusive::list_member_hook<
+  boost::intrusive::link_mode<boost::intrusive::auto_unlink>>;
+
+class Operation;
+class Blocker;
+
+template <typename... T>
+class blocking_future {
+  friend class Operation;
+  friend class Blocker;
+  Blocker *blocker = nullptr;
```

Comment:
nit, s/= nullptr//

Reply:
afaik, types without a constructor have undefined contents if not initialized.  Did that change?

Reply:
Ah, because my only constructor does initialize it.  Will fix.